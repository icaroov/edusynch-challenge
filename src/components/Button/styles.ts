import styled, { css, DefaultTheme } from 'styled-components'
import { darken } from 'polished'

import { ButtonProps } from '.'

export type ContainerButtonProps = Pick<ButtonProps, 'size' | 'fullWidth'>

const containerModifiers = {
  normal: (theme: DefaultTheme) => css`
    height: 4.8rem;
    font-size: ${theme.font.sizes.small};
  `,

  large: (theme: DefaultTheme) => css`
    height: 5.6rem;
    font-size: ${theme.font.sizes.large};
    padding: 1.5rem 7.7rem;
  `,

  fullWidth: () => css`
    width: 100%;
  `
}

export const Container = styled.button<ContainerButtonProps>`
  ${({ theme, size, fullWidth }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;

    border: 0;
    padding: 1.5rem 4.4rem;
    border-radius: ${theme.border.radius};
    font-weight: ${theme.font.bold};

    text-decoration: none;
    color: ${theme.colors.primary};
    background: ${theme.colors.secondary};
    transition: all 0.2s ease;

    &:hover {
      background: ${darken(0.3, theme.colors.tertiary)};
    }

    cursor: pointer;

    ${!!size && containerModifiers[size](theme)};
    ${!!fullWidth && containerModifiers.fullWidth()};
  `}
`
