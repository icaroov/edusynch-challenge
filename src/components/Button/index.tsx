import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'

import * as Styled from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'normal' | 'large'
  fullWidth?: boolean
  as?: React.ElementType
} & ButtonTypes

const Button = ({
  size = 'normal',
  fullWidth = false,
  children,
  ...props
}: ButtonProps) => (
  <Styled.Container size={size} fullWidth={fullWidth} {...props}>
    {!!children && <span>{children}</span>}
  </Styled.Container>
)

export default Button
