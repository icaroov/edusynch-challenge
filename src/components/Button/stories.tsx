import { Story, Meta } from '@storybook/react/types-6-0'

import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button
} as Meta

export const Default: Story<ButtonProps> = args => <Button {...args} />

Default.args = {
  children: 'Get Started'
}

export const asLink: Story = args => <Button {...args} />

asLink.args = {
  children: 'SEARCH',
  as: 'a',
  href: '/link'
}
