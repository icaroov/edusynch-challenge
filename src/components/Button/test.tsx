import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import Button from '.'

describe('<Button />', () => {
  it('should render the button in normal size', () => {
    const { container } = renderWithTheme(<Button>Get Started</Button>)

    expect(screen.getByRole('button', { name: /get started/i })).toHaveStyle({
      height: '4.8rem',
      padding: '1.5rem 4.4rem',
      'font-size': '1.4rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the button with large size', () => {
    renderWithTheme(<Button size="large">Get Started</Button>)

    expect(screen.getByRole('button', { name: /get started/i })).toHaveStyle({
      height: '5.6rem',
      padding: '1.5rem 7.7rem',
      'font-size': '1.8rem'
    })
  })

  it('should render the button with full width', () => {
    renderWithTheme(<Button fullWidth>Get Started</Button>)

    expect(screen.getByRole('button', { name: /get started/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render the button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Get Started
      </Button>
    )

    expect(screen.getByRole('link', { name: /get started/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
