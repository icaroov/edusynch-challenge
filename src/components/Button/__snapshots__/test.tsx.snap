// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the button in normal size 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 0;
  padding: 1.5rem 4.4rem;
  border-radius: 0.4rem;
  font-weight: 700;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #7A57FD;
  background: #FFD74F;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  cursor: pointer;
  height: 4.8rem;
  font-size: 1.4rem;
}

.c0:hover {
  background: #ffc70b;
}

<button
  class="c0"
>
  <span>
    Get Started
  </span>
</button>
`;
